### DISCUSSION TITLE
ATtiny85 20x PGA 1V1 AREF problem

### DISCUSSION THREAD

**Discussion by stevensarns** [SpenceKonde/ATTinyCore#700] [Support / Q & A]
Hello all, thanks in advance for any thoughts.   The code below demonstrates (as succinctly as possible) a problem that I am experiencing with the ATtiny85.  I realize this is a obsolete part and my not be worth the trouble investigating, but there may be an underlying problem that does need correction.  Basically when I measure an analog input with 20x gain and 1.1V Vref the result is zero.  I have not investigated if this is related to SoftwareSerial (I do not have the ability to debug).  The PGA negative input is grounded externally, but declaring the pin an OUTPUT and setting it low also seems to work.  TEMPERATUREand VOLTAGE return reasonable results.

`// ATtiny85 12V gelcell battery charger
#include <SoftwareSerial.h>
SoftwareSerial xSerial(-1,0);   // RX, TX, RX defeated (-1)

// ----------------------- Pins ----------------------------------------
// PB0 TX                         debug data to serial converter on PC
// PB1 PWM                        current / voltage control
// PB2 Vbat                       measure battery terminal voltage
// PB3 ground return diff amp -   diff amp - input
// PB4 current diff amp +         diff amp + to current sense resistor (.02 Ohm)
// PB5 reset

byte const pwmPin   = PIN_PB1;
byte const gndPin   = PIN_PB3;

void setup() {
  xSerial.begin(9600);
  xSerial.println(); 
  xSerial.println(); 
  xSerial.println("Start ");
  xSerial.println();

  pinMode(gndPin, OUTPUT);      // ground return for differential input
  digitalWrite(gndPin, LOW);    // ground
  analogReference(INTERNAL1V1); // Vref = 1.1V
}

byte TEMPERATURE  = B1111;      // mux=1111, temperature
byte CURRENT      = B0111;      // mux=0111, PB4-PB3, 20x gain
byte VOLTAGE      = B0001;      // mux=0001, PB2, 1x
char TAB = 9;
int xxx;

// -------------------------- Main ----------------------------------
void loop() { 
  xxx = analogRead(CURRENT);
  xSerial.print(xxx);
  xSerial.print(TAB);
//  xxx = analogRead(VOLTAGE);
//  xSerial.print(xxx);
//  xSerial.print(TAB);
  xxx = analogRead(TEMPERATURE);
  xSerial.println(xxx);
  delay(1000);
}

//  xxx = 123;                   // works
//  xxx = analogRead(A2);        // works
//  xxx = analogRead(B0010);     // PB4 works (A2)
//  xxx = analogRead(B0110);     // PB4-PB3 works (1x, 1.1Vref, pot connected)
//  xxx = analogRead(B0111);     // PB4-PB3 works (20x, 1.1Vref, pot connected)
//  xxx = analogRead(B0111);     // PB4-PB3, 20x, ref=1.1, works
// measuring CURRENT works
// measuring VOLTAGE & TEMPERATURE works
// measuring CURRENT and anything else causes current = 0, sequence does not matter

`

**Comment by SpenceKonde** [comment]
Differential ADC is not supported in released ATTinyCore. Frankly ADC pin numbering is a godawful mess, and that was the breaking change in 2.0.0-dev that made is absolutely necessary to bump the major version. 2.0.0-dev provides or will provide first-class citizen status to differential ADC measurements, - there's a named constant provided for every ADC differential channel and gain combination; those *must* be used to perform differential readings on 2.0.0-dev, but it works

I wouldn't call the 85 "obsolete" (that term is usually used to designate a part that has ceased production; Microchip continues to make parts far older and less popular than the t85), though it's no spring chicken, and I don't use them. ATTinyCore users still outnumber users of my other cores.

**Comment by SpenceKonde** [comment]
Also, don't use SoftwareSerial - use the builtin software serial implementation, Serial. it is a much more efficient software serial implementation, and in 2.0.0 it will be even more so (I effectively completely reimplemented all the inline assembly, which was deficient in myriad ways), AND it doesn't eat the PCINT interrupt vector, and it's TX pin is by default (switching TX pins is broken in the released version but will work in 2.0.0-dev, though the port is fixed, which doesn't matter on the t85 but does for everything else), switching RX pins is not possible (it will be on the x61's only in 2.0.0, since the analog comparator can compare 3 different pins to the bandgap reference). SoftwareSerial doesn't even use inline assembly; frankly I am amazed that RX works at all, ever, on anything using that library, having studied the code - it's baud rate accuracy is dependent on the compiler version and options. 

```
Serial.begin(9600); 
ACSR &= ~(1<<ACIE); // turn off receive interrupt
ACSR |=  (1<<ACD);    // turn off AC. 
``` 
(it compares the RX line value to the bandgap reference with the analog comparator instead of using the PCINTs, which are broadly useful for user code).

**Comment by stevensarns** [comment]
Wow!  Thank you for the quick reply.  Got it.

**Comment by stevensarns** [comment]
Hello Spence, I am sending this privately because I am kinda embarrassed to
ask.  I wrote this code to directly control the ADC, but the results are
inconsistent.  Can you provide any insights as to what I am doing wrong?  I
have attempted to find your ADC code but was unsuccessful - if you would
provide a link to your corresponding code, you don't have to do anything
else.  Thanks - ss


// -------------------------- ADC -----------------------------
// conversion takes 104 uS, confirmed
// temperature reads 262 at 27C
// ADMUX  REFS1 REFS0 ADLAR REFS2 MUX3  MUX2  MUX1  MUX0
// ADCSRA ADEN  ADSC  ADATE ADIF  ADIE  ADPS2 ADPS1 ADPS0

byte TEMPERATURE  = B1111;      // mux=1111, mux & gain values for desired
input
byte CURRENT      = B0111;      // mux=0111, PB4-PB3, 20x gain
byte VOLTAGE      = B0001;      // mux=0001, PB2, 1x
byte TESTPIN      = B0011;      // mux=0011, PB3, 1x
byte VREF_VCC     = B00000000;  // Vcc is Vref
byte VREF_1V1     = B10000000;  // 1.1v is Vref
byte START_ADC    = B11000110;  // 125 kHz

int analogInput(byte MUX) {     // MUX channel selection input
  int result;
  ADMUX = VREF_1V1 | MUX;       // REF=1v1
//  ADMUX = VREF_VCC | MUX;     // REF=vcc
  ADCSRA = START_ADC;           // begin conversion
  while(bitRead(ADCSRA,ADSC))   // wait for ADSC=0
  result = (ADCH << 8) | ADCL;  // combine bytes
  return result;
}

**Comment by stevensarns** [comment]
So!  Datasheet page 135, Table 17-4, footnote 2 "After switching to internal voltage reference the ADC requires a settling time of 1ms before measurements are stable. Conversions starting before this may not be reliable. The ADC must be enabled during the settling time".  I had ignored this because I was not switching the reference (1.1V).  However the following code works:
`// ------------------------- ADC ------------------------------
// ADCSRA ADEN  ADSC  ADATE ADIF  ADIE  ADPS2 ADPS1 ADPS0
// ADCSRB BIN   ACME  IPR   ----  ----  ADTS2 ADTS1 ADTS0
// ADCMUX REFS1 REFS0 ADLAR REFS2 MUX3  MUX2  MUX1  MUX0
int adcRead(byte chan) {
  int result;
  ADCSRA = B10000110;     // ADEN=1, ADPS=110
  ADMUX  = B10000000;     // ADLAR=0, Vref=010
  ADMUX  = ADMUX | chan;  // or in mux setting
  delay(1);
  bitSet(ADCSRA,ADSC);    // start conversion
  while(ADCSRA & (1<<ADSC));
  result = ADCL | (ADCH<<8);
  return result;
}`
